### PR TITLE
chore(quantic): refactor the Quantic Search Box and Quantic Standalone Search Box components

### PR DESCRIPTION
**Pull request by SimonMilord**
[SFINT-5285](https://coveord.atlassian.net/browse/SFINT-5285)

**ISSUE:**
I want to remove the code duplication present between  the Quantic Search Box and Quantic Standalone Search Box components thus we should refactor these two components and encourage code reuse and optimize the logic that powers these components.

**SOLUTION:**
- Created a seperate component called `quanticSearchBoxInput` which will contain all the logic tied to the `<Input>` or `<textarea`. It supports expandable or not (with different templates) and also support standalone or not.
- Also removed some of the duplication of CSS that was already covered in `searchBoxStyle`

[SFINT-5285]: https://coveord.atlassian.net/browse/SFINT-5285?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/quantic/cypress/e2e/default-2/standalone-search-box/standalone-search-box-expectations.ts [hunk 1]
```diff
@@ -17,9 +17,9 @@ function standaloneSearchBoxExpectations(
         .should(display ? 'exist' : 'not.exist')
         .logDetail(`${should(display)} display the input search box`);
     },
-    inputInitialized: (textarea = false) => {
+    searchBoxInputInitialized: () => {
```
Comment (mmitiche) [inline packages/quantic/cypress/e2e/default-2/standalone-search-box/standalone-search-box-expectations.ts:20]: Why are we removing this? 
in our tests we still wanna confirm that the correct type of input is being used.

File: packages/quantic/cypress/e2e/default-2/standalone-search-box/standalone-search-box-selectors.ts [hunk 2]
```diff
@@ -12,18 +13,28 @@ export interface StandaloneSearchBoxSelector extends ComponentSelector {
...
     ),
   suggestionList: () =>
     StandaloneSearchBoxSelectors.get().find(
-      'c-quantic-search-box-suggestions-list li'
+      'c-quantic-search-box-input c-quantic-search-box-suggestions-list li'
     ),
   clearButton: () =>
-    StandaloneSearchBoxSelectors.get().find('.searchbox__clear-button'),
+    StandaloneSearchBoxSelectors.get().find(
+      'c-quantic-search-box-input button[name="Clear"]'
```
Comment (mmitiche) [inline packages/quantic/cypress/e2e/default-2/standalone-search-box/standalone-search-box-selectors.ts:30]: it's a better practice to use a special attribute to select elements for cypress, we usually use the attribute `data-cy` this attribute only serves for cypress selectors, thus it reduces the chances of accidentally updating the `name` attribute in the HTML of the component breaking by this the current tests.

Comment (mmitiche) [inline packages/quantic/cypress/e2e/default-2/standalone-search-box/standalone-search-box-selectors.ts:30]: it's a better practice to use a special attribute to select elements for cypress, we usually use the attribute `data-cy` this attribute only serves for cypress selectors, thus it reduces the chances of accidentally updating the `name` attribute in the HTML of the component breaking by this the current tests.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 3]
```diff
@@ -120,17 +106,14 @@ export default class QuanticSearchBox extends LightningElement {
...
-      'suggestionlistrender',
-      this.handleSuggestionListEvent
-    );
   }
 
   updateState() {
     if (this.state?.value !== this.searchBox.state.value) {
-      this.input.value = this.searchBox.state.value;
+      // @ts-ignore
+      this.quanticSearchBoxInput.inputValue = this.searchBox.state.value;
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:114]: This line shouldn't be necessary as long as you are passing the `value` as a property to the QuanticSearchBoxInput component like so: 
```
  <c-quantic-search-box-input
    textarea={textarea}
    input={input}
    without-submit-button={withoutSubmitButton}
    placeholder={placeholder}
    search-box-engine={searchBox}
    value={value}
    suggestions={suggestionsArray}
  ></c-quantic-search-box-input>
```

**Review by mmitiche — COMMENTED** [review]
Nice job 👍

Added a comment that would make this solution imo, more scalable and maintainable.

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/templates/searchBox.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<!-- sldsValidatorIgnore -->
```
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/templates/searchBox.html:1]: no need, I think I must have forgotten to remove it

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js [hunk 5]
```diff
@@ -139,189 +122,24 @@ export default class QuanticSearchBox extends LightningElement {
...
-    this.searchBox.submit();
-    this.input.blur();
-  }
-
-  handleKeyValues() {
-    if (this.searchBox?.state?.value !== this.input.value) {
-      this.suggestionList?.resetSelection();
-      this.searchBox.updateText(this.input.value);
-    }
+  get suggestionsArray() {
```
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:139]: I can't use `suggestions` since it is already a duplicate name

image

[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/quanticSearchBox.js:139]: In this case you don't even need this getter, since it's just returning `this.suggestions`

File: packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+      : this.template.querySelector('input');
+  }
+
+  handleEnter() {
+    const selectedSuggestion = this.suggestionList?.getCurrentSelectedValue();
+    if (this.suggestionsOpen && selectedSuggestion) {
+      this.searchBoxEngine.selectSuggestion(selectedSuggestion.rawValue);
+      this.input.blur();
+    } else {
+      this.searchBoxEngine.submit();
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js:137]:
```suggestion
      // emit custom event to submit.
```

File: packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+      this.input.blur();
+    }
+  }
+
+  handleValueChange() {
+    if (this.searchBoxEngine.state.value !== this.input.value) {
+      this.searchBoxEngine.updateText(this.input.value);
+    }
+  }
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js:146]:
```suggestion
  handleValueChange() {
    //emit custom event inputValueChange with the new value
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+  onSubmit(event) {
+    event.stopPropagation();
+    if (this.searchBoxEngine?.state?.value !== this.input.value) {
+      this.searchBoxEngine.updateText(this.input.value);
+    }
+    this.searchBoxEngine.submit();
+    this.input.blur();
+  }
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/quanticSearchBoxInput.js:155]:
```suggestion
  onSubmit(event) {
    event.stopPropagation();
    // emit custom event inputValueChange with the new value
    // emit submit custom event
    this.input.blur();
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/templates/expandableSearchBoxInput.html [hunk 9]
```diff
@@ -1,20 +1,21 @@
-<!-- sldsValidatorIgnore -->
 <template>
   <div class="slds-form-element z-index-high searchbox__container-wrapper">
     <div class="slds-form-element__control searchbox_floating-container">
       <div class="slds-combobox_container">
         <div
           class="slds-combobox slds-dropdown-trigger slds-dropdown-trigger_click"
-          aria-expanded="false"
+          aria-expanded="true"
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticSearchBoxInput/templates/expandableSearchBoxInput.html:7]: This should start false, and it should be updated to true only when the suggestions are visible. This is accessibility requirements.

File: packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/quanticStandaloneSearchBox.js [hunk 10]
```diff
@@ -124,7 +115,7 @@ export default class QuanticStandaloneSearchBox extends NavigationMixin(
   /** @type {Function} */
   unsubscribe;
   /** @type {boolean} */
-  isInitialized;
+  isInitialized = false;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticStandaloneSearchBox/quanticStandaloneSearchBox.js:118]: This is implicit but sure.

File: packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css [hunk 11]
```diff
@@ -3,22 +3,46 @@
 .searchbox__container {
   display: flex;
   position: relative;
-  border-color: #e5e8e8;
   min-height: 48px;
   box-sizing: border-box;
   border-radius: 0.25rem;
-  overflow: hidden;
+  overflow: visible;
+  border: var(--lwc-borderWidthThin, 1px) solid var(--lwc-brandAccessible, #0176d3);
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css:10]: Why are we adding a border to the container? The borders should already be in the inputs etc.

File: packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css [hunk 12]
```diff
@@ -3,22 +3,46 @@
 .searchbox__container {
   display: flex;
   position: relative;
-  border-color: #e5e8e8;
   min-height: 48px;
   box-sizing: border-box;
   border-radius: 0.25rem;
-  overflow: hidden;
+  overflow: visible;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css:9]: Really not sure about this either.

File: packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css [hunk 13]
```diff
@@ -3,22 +3,46 @@
...
 .searchbox__input {
-  display: flex;
   border-radius: 0.25rem;
   border-bottom-left-radius: 0.25rem;
   position: relative;
   height: 48px;
   box-sizing: border-box;
 }
 
+textarea.searchbox__input {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css:35]: This should be in the expandableSearchBoxInput not here
The common style doesn't need to know about the textarea style

File: packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css [hunk 14]
```diff
@@ -27,8 +51,9 @@
 
 .searchbox__submit-button {
   width: 46.75px;
-  border-radius: 0;
   height: 48px;
+  border: none;
+  border-radius: 0 var(--lwc-borderRadiusMedium, 0.25rem) var(--lwc-borderRadiusMedium, 0.25rem) 0;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/searchBoxStyle/searchBoxStyle.css:56]: Not sure about this either, why was this added here? Is it because it was in both css files?

File: packages/quantic/force-app/main/default/lwc/quanticSearchBox/templates/searchBox.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticSearchBox/templates/searchBox.html:1]: I know :) I have yet to push the changes
